src/Disks/ObjectStorages/IObjectStorage.h

```diff
@@ -143,28 +142,35 @@ struct RelativePathWithMetadata
     std::optional<DataFileMetaInfoPtr> file_meta_info;
     /// Retry request after short pause
     CommandInTaskResponse command;
+    std::optional<String> absolute_path;
+    ObjectStoragePtr object_storage_to_use = nullptr;
 
-    RelativePathWithMetadata() = default;
+    PathWithMetadata() = default;
 
-    explicit RelativePathWithMetadata(String command_or_path, std::optional<ObjectMetadata> metadata_ = std::nullopt)
+    explicit PathWithMetadata(
+        const String & command_or_path,
+        std::optional<ObjectMetadata> metadata_ = std::nullopt,
+        std::optional<String> absolute_path_ = std::nullopt,
+        ObjectStoragePtr object_storage_to_use_ = nullptr)
         : relative_path(std::move(command_or_path))
         , metadata(std::move(metadata_))
         , command(relative_path)
+        , absolute_path((absolute_path_.has_value() && !absolute_path_.value().empty()) ? absolute_path_ : std::nullopt)
```

Comment:
If empty is being treated the same as `std::nullopt`, why not just `std::string`?

Reply:
initial idea was to make the object reading logic shorter and avoid a lot of `if (!absolute_path.empty())`
I agree, maybe it shall be refactored. But since this is a more or less style isue, I would prefer to merge it as is now not to delay release any further and improve it when submitting to upstream.